crates/core/tedge_api/src/workflow/state.rs

```diff
@@ -176,39 +239,84 @@ impl GenericCommandState {
     }
 
     pub fn operation(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", operation, _] => Some(operation.to_string()),
-            _ => None,
-        }
+        command_identifier(&self.topic.name).map(|(operation, _)| operation.to_string())
     }
 
     pub fn cmd_id(&self) -> Option<String> {
-        match self.topic.name.split('/').collect::<Vec<&str>>()[..] {
-            [_, _, _, _, _, "cmd", _, cmd_id] => Some(cmd_id.to_string()),
-            _ => None,
-        }
+        command_identifier(&self.topic.name).map(|(_, cmd_id)| cmd_id.to_string())
+    }
+
+    pub fn is_successful(&self) -> bool {
+        matches!(self.status.as_str(), SUCCESSFUL)
+    }
+
+    pub fn is_failed(&self) -> bool {
+        matches!(self.status.as_str(), FAILED)
+    }
+}
+
+/// Return the invoking command topic name, if any
+pub fn invoking_command(sub_command: &TopicName) -> Option<TopicName> {
+    match sub_command.split('/').collect::<Vec<&str>>()[..] {
+        [pre, t1, t2, t3, t4, "cmd", _, sub_id] => extract_invoking_command_id(sub_id)
+            .map(|(op, id)| format!("{pre}/{t1}/{t2}/{t3}/{t4}/cmd/{op}/{id}")),
+        _ => None,
+    }
+}
+
+/// Build a sub command identifier from its invoking command identifier
+///
+/// Using such a structure command id for sub commands is key
+/// to retrieve the invoking command of a sub-command from its state using [extract_invoking_command_id].
+fn sub_command_id(operation: &str, cmd_id: &str) -> String {
+    format!("sub:{operation}:{cmd_id}")
+}
```

Comment:
I'm not happy with this solution to track the invoking command of a sub-command.

Beyond the fact that this doesn't work with sub-sub-commands (this can be fixed), this is polluting the command topic. Somehow using a hidden field in the sub-command state was cleaner.